File: backend/app/infrastructure/models/scm.py
import enum
from sqlalchemy import Column, Integer, String, Boolean, Enum, Float, Date, Text, ForeignKey
from sqlalchemy.orm import relationship
from app.core.database import Base
from app.infrastructure.models.base import TimestampMixin, SoftDeleteMixin


class EstadoSolicitudSCM(str, enum.Enum):
    BORRADOR = "BORRADOR"
    PENDIENTE = "PENDIENTE"
    APROBADA = "APROBADA"
    RECHAZADA = "RECHAZADA"
    EN_PROCESO = "EN_PROCESO"
    COMPLETADA = "COMPLETADA"
    CANCELADA = "CANCELADA"


class PrioridadSCM(str, enum.Enum):
    BAJA = "BAJA"
    MEDIA = "MEDIA"
    ALTA = "ALTA"
    URGENTE = "URGENTE"


class EstadoOrdenSCM(str, enum.Enum):
    BORRADOR = "BORRADOR"
    ENVIADA = "ENVIADA"
    CONFIRMADA = "CONFIRMADA"
    EN_TRANSITO = "EN_TRANSITO"
    RECIBIDA_PARCIAL = "RECIBIDA_PARCIAL"
    RECIBIDA = "RECIBIDA"
    CERRADA = "CERRADA"
    CANCELADA = "CANCELADA"


class CategoriaSCM(str, enum.Enum):
    INSUMOS = "INSUMOS"
    SERVICIOS = "SERVICIOS"
    EQUIPOS = "EQUIPOS"
    MATERIALES = "MATERIALES"
    LOGISTICA = "LOGISTICA"
    IT = "IT"
    REPUESTOS = "REPUESTOS"
    PAPELERIA = "PAPELERIA"
    OTROS = "OTROS"


class ClasificacionProveedor(str, enum.Enum):
    A = "A"
    B = "B"
    C = "C"
    D = "D"


class RecomendacionProveedor(str, enum.Enum):
    MANTENER = "MANTENER"
    MEJORAR = "MEJORAR"
    REEMPLAZAR = "REEMPLAZAR"
    SUSPENDER = "SUSPENDER"


class ScmSolicitudCompra(Base, TimestampMixin, SoftDeleteMixin):
    __tablename__ = "scm_solicitudes_compra"
    __table_args__ = {"extend_existing": True}

    id = Column(Integer, primary_key=True, index=True)
    numero = Column(String(50), unique=True, nullable=False, index=True)
    solicitante_id = Column(Integer, ForeignKey("usuarios.id"), nullable=False)
    aprobador_id = Column(Integer, ForeignKey("usuarios.id"), nullable=True)
    proveedor_id = Column(Integer, ForeignKey("proveedores.id"), nullable=True)
    titulo = Column(String(300), nullable=False)
    descripcion = Column(Text, nullable=True)
    categoria = Column(Enum(CategoriaSCM), nullable=False, default=CategoriaSCM.OTROS)
    prioridad = Column(Enum(PrioridadSCM), nullable=False, default=PrioridadSCM.MEDIA)
    estado = Column(Enum(EstadoSolicitudSCM), nullable=False, default=EstadoSolicitudSCM.BORRADOR)
    fecha_requerida = Column(Date, nullable=True)
    presupuesto_estimado = Column(Float, nullable=True)
    moneda = Column(String(10), default="COP")
    justificacion = Column(Text, nullable=True)
    observaciones = Column(Text, nullable=True)
    fecha_aprobacion = Column(Date, nullable=True)
    motivo_rechazo = Column(Text, nullable=True)

    items = relationship("ScmSolicitudItem", back_populates="solicitud", cascade="all, delete-orphan")
    ordenes = relationship("ScmOrdenCompra", back_populates="solicitud")


class ScmSolicitudItem(Base, TimestampMixin):
    __tablename__ = "scm_solicitud_items"
    __table_args__ = {"extend_existing": True}

    id = Column(Integer, primary_key=True, index=True)
    solicitud_id = Column(Integer, ForeignKey("scm_solicitudes_compra.id"), nullable=False)
    descripcion = Column(String(500), nullable=False)
    unidad = Column(String(50), nullable=True)
    cantidad = Column(Float, nullable=False, default=1)
    precio_estimado = Column(Float, nullable=True)
    total_estimado = Column(Float, nullable=True)
    especificaciones = Column(Text, nullable=True)

    solicitud = relationship("ScmSolicitudCompra", back_populates="items")


class ScmOrdenCompra(Base, TimestampMixin, SoftDeleteMixin):
    __tablename__ = "scm_ordenes_compra"
    __table_args__ = {"extend_existing": True}

    id = Column(Integer, primary_key=True, index=True)
    numero = Column(String(50), unique=True, nullable=False, index=True)
    solicitud_id = Column(Integer, ForeignKey("scm_solicitudes_compra.id"), nullable=True)
    proveedor_id = Column(Integer, ForeignKey("proveedores.id"), nullable=False)
    creado_por_id = Column(Integer, ForeignKey("usuarios.id"), nullable=False)
    aprobado_por_id = Column(Integer, ForeignKey("usuarios.id"), nullable=True)
    estado = Column(Enum(EstadoOrdenSCM), nullable=False, default=EstadoOrdenSCM.BORRADOR)
    categoria = Column(Enum(CategoriaSCM), nullable=False, default=CategoriaSCM.OTROS)
    prioridad = Column(Enum(PrioridadSCM), nullable=False, default=PrioridadSCM.MEDIA)
    fecha_emision = Column(Date, nullable=True)
    fecha_entrega_esperada = Column(Date, nullable=True)
    fecha_entrega_real = Column(Date, nullable=True)
    subtotal = Column(Float, nullable=True, default=0)
    impuestos = Column(Float, nullable=True, default=0)
    total = Column(Float, nullable=True, default=0)
    moneda = Column(String(10), default="COP")
    condiciones_pago = Column(String(200), nullable=True)
    lugar_entrega = Column(String(300), nullable=True)
    notas = Column(Text, nullable=True)
    codigo_sap = Column(String(50), nullable=True)

    items = relationship("ScmOrdenItem", back_populates="orden", cascade="all, delete-orphan")
    solicitud = relationship("ScmSolicitudCompra", back_populates="ordenes")


class ScmOrdenItem(Base, TimestampMixin):
    __tablename__ = "scm_orden_items"
    __table_args__ = {"extend_existing": True}

    id = Column(Integer, primary_key=True, index=True)
    orden_id = Column(Integer, ForeignKey("scm_ordenes_compra.id"), nullable=False)
    descripcion = Column(String(500), nullable=False)
    codigo_producto = Column(String(100), nullable=True)
    unidad = Column(String(50), nullable=True)
    cantidad = Column(Float, nullable=False, default=1)
    cantidad_recibida = Column(Float, nullable=True, default=0)
    precio_unitario = Column(Float, nullable=False, default=0)
    descuento_pct = Column(Float, nullable=True, default=0)
    total = Column(Float, nullable=True, default=0)
    especificaciones = Column(Text, nullable=True)

    orden = relationship("ScmOrdenCompra", back_populates="items")


class ScmEvaluacionProveedor(Base, TimestampMixin):
    __tablename__ = "scm_evaluaciones_proveedor"
    __table_args__ = {"extend_existing": True}

    id = Column(Integer, primary_key=True, index=True)
    proveedor_id = Column(Integer, ForeignKey("proveedores.id"), nullable=False)
    evaluador_id = Column(Integer, ForeignKey("usuarios.id"), nullable=False)
    periodo = Column(String(20), nullable=False)
    calidad = Column(Float, nullable=True)
    tiempo_entrega = Column(Float, nullable=True)
    precio = Column(Float, nullable=True)
    servicio = Column(Float, nullable=True)
    documentacion = Column(Float, nullable=True)
    puntaje_total = Column(Float, nullable=True)
    clasificacion = Column(Enum(ClasificacionProveedor), nullable=True)
    comentarios = Column(Text, nullable=True)
    recomendacion = Column(Enum(RecomendacionProveedor), nullable=True)

File: frontend/src/api/scm.ts
import { apiClient } from './client'

// ─── Enums ────────────────────────────────────────────────────────────────────
export type EstadoSolicitud = 'BORRADOR' | 'PENDIENTE' | 'APROBADA' | 'RECHAZADA' | 'EN_PROCESO' | 'COMPLETADA' | 'CANCELADA'
export type PrioridadSCM   = 'BAJA' | 'MEDIA' | 'ALTA' | 'URGENTE'
export type EstadoOrden    = 'BORRADOR' | 'ENVIADA' | 'CONFIRMADA' | 'EN_TRANSITO' | 'RECIBIDA_PARCIAL' | 'RECIBIDA' | 'CERRADA' | 'CANCELADA'
export type CategoriaSCM   = 'INSUMOS' | 'SERVICIOS' | 'EQUIPOS' | 'MATERIALES' | 'LOGISTICA' | 'IT' | 'REPUESTOS' | 'PAPELERIA' | 'OTROS'
export type ClasificacionProveedor = 'A' | 'B' | 'C' | 'D'

// ─── Interfaces ───────────────────────────────────────────────────────────────
export interface SolicitudItem {
  id?: number
  descripcion: string
  cantidad: number
  unidad?: string
  precio_estimado?: number
  total_estimado?: number
  especificaciones?: string
}

export interface Solicitud {
  id: number
  numero: string
  titulo: string
  descripcion?: string
  estado: EstadoSolicitud
  prioridad: PrioridadSCM
  categoria: CategoriaSCM
  presupuesto_estimado?: number
  fecha_requerida?: string
  fecha_aprobacion?: string
  motivo_rechazo?: string
  items: SolicitudItem[]
  created_at: string
}

export interface OrdenItem {
  id?: number
  descripcion: string
  cantidad: number
  unidad?: string
  precio_unitario: number
  subtotal?: number
  especificaciones?: string
}

export interface OrdenCompra {
  id: number
  numero: string
  proveedor_id: number
  proveedor_nombre?: string
  solicitud_id?: number
  estado: EstadoOrden
  prioridad: PrioridadSCM
  subtotal: number
  impuestos: number
  total: number
  fecha_entrega_esperada?: string
  fecha_entrega_real?: string
  condiciones_pago?: string
  notas?: string
  items: OrdenItem[]
  created_at: string
}

export interface EvaluacionProveedor {
  id: number
  proveedor_id: number
  orden_compra_id?: number
  calidad: number
  tiempo_entrega: number
  precio: number
  servicio: number
  documentacion: number
  puntaje_total: number
  clasificacion: ClasificacionProveedor
  comentarios?: string
  recomendacion?: string
  created_at: string
}

export interface ProveedorSCM {
  id: number
  razon_social: string
  nombre_comercial?: string
  nit?: string
  contacto_nombre?: string
  contacto_email?: string
  ciudad?: string
  tipo?: string
  total_ordenes?: number
  puntaje_promedio?: number
  clasificacion?: ClasificacionProveedor
}

// Dashboard usa dicts para los estados (formato real del backend)
export interface SCMDashboardData {
  kpis: {
    total_solicitudes: number
    solicitudes_pendientes: number
    oc_abiertas: number
    valor_oc_en_proceso: number
    proveedores_activos: number
  }
  oc_por_estado: Record<string, number>
  sol_por_estado: Record<string, number>
}

// ─── API calls ────────────────────────────────────────────────────────────────
export async function getSCMDashboard(): Promise<SCMDashboardData> {
  const r = await apiClient.get('/scm/dashboard')
  return r.data
}

export async function getSolicitudes(params?: {
  page?: number; page_size?: number; estado?: EstadoSolicitud; prioridad?: PrioridadSCM
}): Promise<{ items: Solicitud[]; total: number }> {
  const r = await apiClient.get('/scm/solicitudes', { params })
  return r.data
}

export async function createSolicitud(data: {
  titulo: string; descripcion?: string; prioridad: PrioridadSCM; categoria: CategoriaSCM
  presupuesto_estimado?: number; fecha_requerida?: string; items: SolicitudItem[]
}): Promise<Solicitud> {
  const r = await apiClient.post('/scm/solicitudes', data)
  return r.data
}

export async function enviarSolicitud(id: number): Promise<Solicitud> {
  const r = await apiClient.put(`/scm/solicitudes/${id}/enviar`)
  return r.data
}

export async function aprobarSolicitud(id: number): Promise<Solicitud> {
  const r = await apiClient.put(`/scm/solicitudes/${id}/aprobar`)
  return r.data
}

export async function rechazarSolicitud(id: number, motivo: string): Promise<Solicitud> {
  const r = await apiClient.put(`/scm/solicitudes/${id}/rechazar`, { motivo })
  return r.data
}

export async function getOrdenesCompra(params?: {
  page?: number; page_size?: number; estado?: EstadoOrden; proveedor_id?: number
}): Promise<{ items: OrdenCompra[]; total: number }> {
  const r = await apiClient.get('/scm/ordenes-compra', { params })
  return r.data
}

export async function getOrdenCompra(id: number): Promise<OrdenCompra> {
  const r = await apiClient.get(`/scm/ordenes-compra/${id}`)
  return r.data
}

export async function createOrdenCompra(data: {
  proveedor_id: number; solicitud_id?: number; prioridad?: PrioridadSCM
  fecha_entrega_estimada?: string; terminos_pago?: string; notas?: string
  impuestos_pct?: number; items: OrdenItem[]
}): Promise<OrdenCompra> {
  const r = await apiClient.post('/scm/ordenes-compra', data)
  return r.data
}

export async function actualizarEstadoOrden(id: number, estado: EstadoOrden, fecha_entrega_real?: string): Promise<OrdenCompra> {
  const r = await apiClient.put(`/scm/ordenes-compra/${id}/estado`, { estado, fecha_entrega_real })
  return r.data
}

export async function getEvaluacionesProveedor(proveedorId: number): Promise<EvaluacionProveedor[]> {
  const r = await apiClient.get(`/scm/evaluaciones/proveedor/${proveedorId}`)
  return r.data
}

export async function createEvaluacion(data: {
  proveedor_id: number; orden_compra_id?: number
  calidad: number; tiempo_entrega: number; precio: number; servicio: number; documentacion: number
  comentarios?: string; recomendacion?: string
}): Promise<EvaluacionProveedor> {
  const r = await apiClient.post('/scm/evaluaciones', data)
  return r.data
}

export async function getProveedoresSCM(params?: {
  page?: number; page_size?: number; q?: string
}): Promise<{ items: ProveedorSCM[]; total: number }> {
  const r = await apiClient.get('/scm/proveedores', { params })
  return r.data
}
